### ISSUE TITLE
Better tools for anti-harassment

### ISSUE THREAD

**Issue by krues8dr** [opengovfoundation/madison#920]
We've been talking about incremental changes to our system to deal with harassment and trolling via comments for a while.  I want us to move on as many of these as possible immediately, so I'm opening up the conversation here to the tools & strategies we want to implement.  Here are the first few I think that we should roll out:

* [ ] Allow document sponsors to hide flagged comments.  These comments will only be hidden by default to prevent censorship, and users can still show them by clicking something to show them (per thread *or* once for the whole doc).
* [ ] Send email notifications to document sponsors and admins on flagged comments.  Show a list of flagged comments to admin users.
* [ ] Send email notifications to document sponsors for new comments an replies.
* [ ] Send email notifications for all replies to comments - but allow the user to unsubscribe from a single thread or a whole document.
* [ ] Allow users to disable email notifications for any/all of these.
* [ ] Allow the original creator of a thread (or an admin or sponsor) to close it from further replies.
* [ ] Allow users to ignore other users.  We may want to create a user page for each user to make this easier.
* [ ] Allow users to flag other users.
* [ ] Allow admins to ban users.
* [ ] Allow admins to hide all comments from a given user.
* [ ] Fix flags & likes on comment replies ( #771 )

@cmbirk @seamuskraft @sethetter : What do you all think?

**Comment by krues8dr** [comment]
Pinging @mattbailey0 @emanuelfeld for feedback here as well.  What tools would help you all in the case of harassment or trolling?

**Comment by cmbirk** [comment]
I would only allow an admin or sponsor to close a thread.  I think it's important that someone commenting knows there's going to be a discussion ( allowed ) on their comment / annotation.

I still hesitate when thinking about banning users, especially in an official capacity.  The idea of hiding low quality comments / users with the ability to discover seems like a better option.

**Comment by emanuelfeld** [comment]
What's the most highly used instance of Madison and what sort of trolling/harassment problems have they faced in comments?

**Comment by krues8dr** [comment]
@cmbirk : The issue you get into is where trolls can brigade on a thread and the user who started it loses control.  This happens on reddit pretty regularly.  If the OP can shut it down, the users have more control over the situation.  And the idea here is to empower users.

@emanuelfeld : We haven't seen any so far, but I'm trying to get us out ahead of the situation that's currently unfolding in other communities.  Of note, the Genius community is on fire right now, and the rest of the annotation world is feeling the blowback from that.  Context: http://genius.it/8890953/www.slate.com/articles/double_x/doublex/2016/03/news_genius_wants_to_annotate_the_entire_web_at_what_cost.html

**Comment by cmbirk** [comment]
@krues8dr I get that, but we've discussed previously how Madison in general isn't conducive to the same user behavior as a general comment thread / Reddit environment.  I would recommend just allowing admins unless you're seeing that as a particular issue.  At this juncture I think it would be best to focus on open conversation.

**Comment by krues8dr** [comment]
@cmbirk I don't see why.  The potential negatives here are that someone could close a thread before everyone has had a chance to weigh in - but those people could always just start a new comment thread.  The potential positives here are that someone could stop brigading without admin intervention.  Weighing the two, I'm much more inclined to proceed with the change than to leave it as-is.

**Comment by rebeccawilliams** [comment]
Q: Will Madison always be used for gov documents only?

Will these tools let the gov erase comments that they find harassing to their gov?

**Comment by cmbirk** [comment]
If someone can't reply to the comment in question, they're likely to just not comment at all.  There's no way to capture that if it's happening.

On the other hand, if you _are_ getting a brigading situation you would be aware that it's happening and able to resolve it.

Just keep in mind that we're trying to attract an audience that isn't necessarily as familiar as we are with the web and general UI patterns.  I can see a very real situation where many less web-inclinded individuals would get frustrated with not being able to reply and just click off to something else rather than taking the time to make the same selection and annotate in response to a closed comment.

**Comment by krues8dr** [comment]
@rebeccawilliams 

>Will Madison always be used for gov documents only?

No, there are already policy documents by non-government organizations being submitted.

>Will these tools let the gov erase comments that they find harassing to their gov?

We've [discussed this at length](https://github.com/opengovfoundation/madison/issues/122), and I think we believe that the gov should only be able to hide comments from the default view, not remove them entirely - but these can be viewed by anyone by clicking something to reveal them.  Similar to how reddit has a threshold that comments ranked below do not show up unless you click to see them.  

Open to other suggestions here.

**Comment by krues8dr** [comment]
@cmbirk we now have a "closed for comments" piece in place on documents themselves, so this isn't an entirely new paradigm or pattern for Madison.  Messaging could easily solve your concerns here.

**Comment by cmbirk** [comment]
Correct, and the 'closed for comments' is specifically controlled by an admin / sponsor and applies to the entire document.

I think there's more to be lost with this solution than gained, especially since we haven't seen this problem at all to date.

**Comment by krues8dr** [comment]
@cmbirk to that end - we, personally, haven't seen *any* of these issues to date.  That, however, is absolutely no reason for us not to take harassment seriously and begin building a series of policies and tools on how to handle it.  Just because the dam hasn't broke is a poor reason not to build levies.  

I'm ok with tabling that particular aspect of this, of course, which is why I brought it up. :)

**Comment by cmbirk** [comment]
I'm good with that.

Another use case to consider is someone who comes to a comment via a social share.  They may be there for the first time and unfamiliar with Madison.  If they don't have an easy way to respond, we can easily lose them without having any idea that it's happening.

The difference I see with Madison vs. Twitter / Reddit / etc is that we're asking people who aren't necessarily web-savvy to contribute online in an unfamiliar environment.  There's obviously still the risk of trolls, but a much larger barrier for adoption for our target audience.

**Comment by shepazu** [comment]
All of this is great from a single-service perspective, and I'd like to see if we can also think about it from a decentralized curation point of view.
